tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -398,14 +412,119 @@ export class WebGPUBackend extends KernelBackend {
   // https://github.com/tensorflow/tfjs/issues/1595
   override readSync(dataId: object): BackendValues {
     const tensorData = this.tensorMap.get(dataId);
-    const {values} = tensorData;
+    const {values, complexTensorInfos} = tensorData;
 
-    if (values == null) {
-      throw new Error(
-          'WebGPU readSync is only available for CPU-resident tensors.');
+    if (values != null || tensorData.dtype === 'string') {
+      return values;
     }
 
-    return values;
+    if (tensorData.dtype === 'complex64') {
+      const realValues =
+          this.readSync(complexTensorInfos.real.dataId) as Float32Array;
+      const imagValues =
+          this.readSync(complexTensorInfos.imag.dataId) as Float32Array;
+      const complexVals = util.convertBackendValuesAndArrayBuffer(
+          backend_util.mergeRealAndImagArrays(realValues, imagValues).buffer,
+          'float32');
+      this.convertAndCacheOnCPU(dataId, complexVals);
+      return complexVals;
+    }
+
+    let channelMask: Partial<Record<GPUCanvasAlphaMode, number>>;
+    if (this.littleEndian) {
+      channelMask = {
+        opaque: 0x00ffffff,
+        premultiplied: 0xff000000,
+      };
+    } else {
+      channelMask = {
+        opaque: 0xffffff00,
+        premultiplied: 0x000000ff,
+      };
+    }
+
+    const bufferInfo = tensorData.resourceInfo as BufferInfo;
+    const size = bufferInfo.size;
+    util.assert(size % 4 === 0, () => 'GPU buffer size must be multiple of 4.');
+    const $size = size / 4;
+    const valsGPU = new ArrayBuffer(size);
+    const side = 256;
+    const stagingDeviceStorage: OffscreenCanvas[] =
```

Comment:
Should we check `typeof OffscreenCanvas !== 'undefined'` before using `OffscreenCanvas`?